File: src/main/java/com/github/stokito/IdeaSingletonInspection/SingletonPatternInspection.java
package com.github.stokito.IdeaSingletonInspection;

import com.intellij.codeInspection.BaseJavaLocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.codeInspection.ui.MultipleCheckboxOptionsPanel;
import com.intellij.psi.PsiElementVisitor;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public class SingletonPatternInspection extends BaseJavaLocalInspectionTool {

  @SuppressWarnings({"PublicField"})
  public boolean m_checkFinal = false;

  @NotNull
  @Override
  public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, final boolean isOnTheFly) {
    return new SingletonPatternInspectionVisitor(holder, m_checkFinal);
  }

  @Override
  public JComponent createOptionsPanel() {
    final MultipleCheckboxOptionsPanel optionsPanel = new MultipleCheckboxOptionsPanel(this);
    optionsPanel.addCheckbox("Check that singletons are final", "m_checkFinal");
    return optionsPanel;
  }

}

File: src/main/java/com/github/stokito/IdeaSingletonInspection/SingletonPatternInspection.kt
package com.github.stokito.IdeaSingletonInspection

import com.intellij.codeInspection.AbstractBaseJavaLocalInspectionTool
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.codeInspection.ui.MultipleCheckboxOptionsPanel
import com.intellij.psi.PsiElementVisitor

import javax.swing.*

class SingletonPatternInspection : AbstractBaseJavaLocalInspectionTool() {

    var m_checkFinal = false

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        return SingletonPatternInspectionVisitor(holder, m_checkFinal)
    }

    override fun createOptionsPanel(): JComponent? {
        val optionsPanel = MultipleCheckboxOptionsPanel(this)
        optionsPanel.addCheckbox("Check that singletons are final", "m_checkFinal")
        return optionsPanel
    }

}
